README
NAME
    intro.pod - introduction to WordNet::SenseRelate::TargetWord

CONTENTS
    *intro.pod*, *install.pod*, *utils.pod*, *modules.pod*,
    *developers.pod*, *config.pod*, FDL.txt

SYNOPSIS
    You can use pod2html, pod2latex, pod2man, or pod2text to get this
    documentation in a different format. See the man pages for pod2html,
    etc. These translators should come with Perl, but you can also download
    them from <http://search.cpan.org>.

DESCRIPTION
    This package consists of a set of Perl modules along with supporting
    Perl programs that perform the task of Word Sense Disambiguation. The
    program(s) attempt to disambiguate the sense of a single target word in
    a given context as described by Banerjee and Pedersen (2002), Patwardhan
    et al. (2003) and Banerjee and Pedersen (2003). This package separates
    the disambiguation process into the following steps:

    (a) Preprocessing

    (b) Context Selection

    (c) Postprocessing

    (d) Sense Selection Algorithm

    The context, which is typically 3 to 4 sentences of text, is passed
    through each of these stages. Each stage does some processing of the
    context. The final sense selection stage picks a sense of the target
    word, using the information from the context. For example, in order to
    determine which sense of a given word is being used in a particular
    context, the sense having the highest relatedness with its context word
    senses is most likely to be the sense being used.

    Because of the wide range of possibilites for each of the stages, the
    package has been made highly configurable. The user is allowed to choose
    different modules for each of the stages. Additionally, various
    configuration options can be passed to the substages to control their
    behavior.

    The WordNet::SenseRelate::TargetWord initializes the modules specified
    by the user for each substage. It then accepts instances of text,
    consisting of a few sentences, with one of the words marked as the
    target word to be disambiguated. It passes the text through each of the
    stages, and gets the chosen sense from the last (Algorithm) stage, and
    return the chosen sense as the answer for that instance. The substages
    are applied in the following order:

    Preprocessing
        WordNet::SenseRelate::TargetWord allows the user to apply multiple
        preprocessing modules to the input text. All of the preprocessing
        modules specified by the user are applied to the text in the order
        specified by the user.

    Context Selection
        The WordNet::SenseRelate::TargetWord package allows the user to
        select one *Context Selection* package, which when applied to the
        context would pick a subset of the words around the target word for
        the disambiguation algorithm to use in picking the intended sense of
        the target word.

    Postprocessing
        Like the preprocessing modules the user can select multiple
        postprocessing modules that would be applied to the text after the
        context selection has been done. Before the postprocessing modules
        are called the program determines the possible senses of the target
        word and the selected context words. The postprocessing stages are
        intended for any processing that involves the senses of the context
        words. For example, applying some heutistics to "prune" some of the
        senses of the target word, or the context words.

    Sense Selection Algorithm
        The goal of the sense selection algorithm is to use the senses of
        the context words to pick one of the senses of the target word as
        the answer. The WordNet::SenseRelate::TargetWord module allows the
        user to specify only one Algorithm module.

INSTALL
    The package can be installed with the following four commands:

     perl Makefile.PL

     make

     make test

     make install

    For a more detailed explanation or for non-standard installations, see
    install.pod.

MODULES
    In this section, we list the modules provided in the package.

    Preprocessing modules
        Currently, only one preprocessing module is provided in the package:

        WordNet::SenseRelate::Preprocess::Compounds
            This module detects collocations within the text, and joins them
            with underscores. For example, a piece of text such as "the
            board of directors" would become "the board_of_directors".

    Context Selection modules
        One context selection module is provided in the package:

        WordNet::SenseRelate::Context::NearestWords
            This modules picks the N words nearest the target word as the
            context words that should be used by the disambiguation
            algorithm. A stop list can be specified to omit words like "a",
            "an, "the", etc. The value N can be set during initialization.

    Postprocessing modules
        As of this vesion, there are no postprocessing modules in the
        package. However, we intend to add some in future releases.

    Sense Selection Algorithm
        Four sense selection algorithm modules are provided in the pacakge:

        WordNet::SenseRelate::Algorithm::Local
            This modules selects that sense of the target word which is most
            related to the senses of the context words. To do this it uses
            the "Local" disambiguation algorithm as described by Banerjee
            and Pedersen (2002). In order to determine the relatedness of
            senses, the algorithm uses one of the WordNet::Similarity
            measures of relatedness. Using the configuration options for
            this module, the user can specify which measure the algorithm
            should use.

        WordNet::SenseRelate::Algorithm::Global
            This modules selects that sense of the target word which is most
            related to the senses of the context words. To do this it uses
            the "Global" disambiguation algorithm as described by Banerjee
            and Pedersen (2002). This algorithm is somewhat similar to the
            "Local" algorithm. It differs from the "Local" algorithm, in
            that it forms all possible combinations of the senses of the
            context words, and evaluates the semantic relatedness for each
            combination separately. The combination with the maximum score
            is is selected, and the sense of the target word in that
            combination is returned as the answer.

        WordNet::SenseRelate::Algorithm::SenseOne
            This module provides a baseline for the disambiguation process
            by always returning the first sense of the target word as the
            answer.

        WordNet::SenseRelate::Algorithm::Random
            This module provides another baseline by randomly selecting one
            of the senses of the target word as the answer.

    Apart from all of the modules mentioned above that form the pieces of
    the bigger structure, the package also contains the
    WordNet::SenseRelate::TargetWord module which combines the above pieces.

    In order to use these modules in a Perl program for Word Sense
    Disambiguation, we need only create an instance of the
    WordNet::SenseRelate::TargetWord module in our program, and provide it
    with options that indicate which of the above modules (along with
    configuration options) it should use in the disambiguation process.

    Additionally, in order to be able to use this package for disambiguating
    instances from the Senseval2 or the Senseval3 data sets, the
    WordNet::SenseRelate::Reader::Senseval2 module has also been provided in
    the package. The reader module reads in an entire Senseval2 formatted
    XML file and builds a list of instances from the file. A Perl program
    can then iterate over these instances and pass them to the
    WordNet::SenseRelate::TargetWord object.

UTILITIES
    Some Perl utilities that provide command-line and graphical interfaces
    to the Perl modules are provided in the '/utils' directory of the
    package.

    disamb.pl
        Performs Word Sense Disambiguation on Senseval-2 lexical sample
        data. It uses the WordNet::SenseRelate::Reader::Senseval2 module to
        read a Senseval2 lexical sample file, and then disambiguates each of
        the instances using the WordNet::SenseRelate::TargetWord module.

        Usage: disamb.pl [ [--config FILE] [--wnpath WNPATH] [--trace] XMLFILE | --help | --version]

        --config=FILENAME  Specifies a configuration file (FILENAME) to set
                           up the various configuration options.

        --wnpath=WNPATH    WNPATH specifies the path of the WordNet data
                           files. Ordinarily, this path is determined from
                           the $WNHOME environment variable. But this option
                           overides this behavior.

        --trace            Indicates that trace information be printed.

        --help             Displays this help screen.

        --version          Displays version information.

        Example:

        To disambiguate an English lexical sample file using the default
        options

          disamb.pl eng-lex-samp.xml

        To dismabiguate an English lexical sample file, specifying
        configuration options and trace output

          disamb.pl --config config.txt --trace eng-lex-sample.xml

    wps2sk.pl
        Creates a word#pos#sense to sensekey mapping of a Senseval-2 answer
        file (output by disamb.pl). In order to be able to evaluate the
        output of disamb.pl using software provided by the Senseval
        organizers, we need to covert the output of disamb.pl to the
        "SenseKey" format.

        Usage: wps2sk.pl [ [ --wnpath WNPATH] [FILE...] | --help | --version ]

        --wnpath=WNPATH   WNPATH specifies the path of the WordNet data
                          files. Ordinarily, this path is determined from
                          the $WNHOME environment variable. But this option
                          overides this behavior.

        --quiet           Run in quiet mode -- does not print informational
                          messages. But it does print warning or error messages
                          if any.

        --help            Displays this help screen.

        --version         Displays version information.

        Example:

        To convert a captured output from disamb.pl

          wps2sk.pl output.txt

        Typically used in a pipe with disamb.pl

          disamb.pl xmlfile.xml | wps2sk.pl --quiet

    disamb-gui.pl
        Performs Word Sense Disambiguation on Senseval-2 lexical sample data
        (graphical interface to WordNet::SenseRelate::TargetWord). It uses
        the WordNet::SenseRelate::Reader::Senseval2 module to read a
        Senseval2 lexical sample file, and then allows the user to pick the
        instances that she wishes to disambiguate.

        Usage: disamb-gui.pl [ [--config FILE] [--wnpath WNPATH] [XMLFILE] | --help | --version]

        --config=FILENAME   Specifies a configuration file (FILENAME) to set
                            up the various configuration options.

        --wnpath=WNPATH     WNPATH specifies the path of the WordNet data
                            files. Ordinarily, this path is determined from
                            the $WNHOME environment variable. But this option
                            overides this behavior.

        --help              Displays this help screen.

        --version           Displays version information.

CONFIGURATION FILE
    A configuration file can be passed to both, disamb.pl as well as to
    disamb-gui.pl in order to set up the configuration parameters for the
    modules. The format of this file is specified in this section.

    The configuration file must start with a header, which is the string
    "WordNet::SenseRelate::TargetWord" on the first byte of the first line
    of the file.

    It consists of a number of sections. Each section contains a list of
    modules for that section, along with the configuration options for the
    modules.

    A section starts with the string

      [SECTION:section_name]

    A new section header indicates the end of the previous section. The
    utilies in this package recognize four types of sections:

      [SECTION:PREPROCESS]

      [SECTION:CONTEXT]

      [SECTION:POSTPROCESS]

      [SECTION:ALGORITHM]

    Any other section will be ignored by the program, without warning. Each
    of the sections is used to specify the list of modules for that section.
    A module is specified by [START modname] and [END] tags appearing on
    separate lines.

    Configuration parameters for the module appear on separate lines in
    between the start and end tags as PARAMETER=VALUE pairs.

    Example config file:

      WordNet::SenseRelate::TargetWord

      [SECTION:PREPROCESS]

      [START WordNet::SenseRelate::Preprocess::Compounds]

      [END]

      [SECTION:CONTEXT]

      [START WordNet::SenseRelate::Context::NearestWords]

      windowsize=5

      windowstop=/home/sid/window-stop.txt

      contextpos=n

      [END]

      [SECTION:POSTPROCESS]

      [SECTION:ALGORITHM]

      [START WordNet::SenseRelate::Algorithm::Local]

      measure=WordNet::Similarity::vector

      [END]

    The above config file has four sections -- PREPROCESS, CONTEXT,
    POSTPROCESS and ALGORITHM. The PREPROCESS section contains a single
    module for compound detection, and has no config parameters. The CONTEXT
    section lists one module (NearestWords), with parameters *windowsize*
    and *windowstop*. There are no modules listed in the POSTPROCESS
    section. The ALGORITHM section lists the Local disambiguation algorithm
    module, also with the *measure* specified.

SOFTWARE COPYRIGHT AND LICENSE
    Copyright (C) 2005 Siddharth Patwardhan, Ted Pedersen, and Satanjeev
    Banerjee

    This suite of programs is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public License as published
    by the Free Software Foundation; either version 2 of the License, or (at
    your option) any later version.

    This program is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
    Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

    Note: The text of the GNU General Public License is provided in the file
    'GPL.txt' that you should have received with this distribution.

ACKNOWLEDGEMENTS
    This research is partially supported by a National Science Foundation
    Faculty Early CAREER Development Award (#0092784).

REFERENCES
    1   S. Banerjee and T. Pedersen. An Adapted Lesk Algorithm for Word
        Sense Disambiguation Using WordNet. In Proceedings of the Fourth
        International Conference on Computational Linguistics and
        Intelligent Text Processing (CICLING-02). Mexico City, Mexico. 2002.

    2   S. Banerjee and T. Pedersen. Extended gloss overlaps as a measure of
        semantic relatedness. In Proceedings of the Eighteenth International
        Conference on Artificial Intelligence (IJCAI-03). Acapulco, Mexico.
        2003.

    3   S. Patwardhan, S. Banerjee and T. Pedersen. Using Semantic
        Relatedness for Word Sense Disambiguation. In Proceedings of the
        Fourth International Conference on Intelligent Text Processing and
        Computational Linguistics (CiCLING-03). Mexico City, Mexico. 2003.

    4   S. Banerjee. Adapting the Lesk algorithm for Word Sense
        Disambiguation to WordNet. Masters Thesis, University of Minnesota,
        Duluth. 2002.

    5   S. Patwardhan. Incorporating Dictionary and Corpus Information into
        a Context Vector Measure of Semantic Relatedness. Masters Thesis,
        University of Minnesota, Duluth. 2003.

    6   Fellbaum C., editor. WordNet: An electronic lexical database. MIT
        Press. 1998.

SEE ALSO
    <http://groups.yahoo.com/group/senserelate>

    <http://search.cpan.org/dist/WordNet-SenseRelate-TargetWord>

    <http://senserelate.sourceforge.net>

AUTHORS
     Siddharth Patwardhan, University of Utah
     sidd at cs.utah.edu

     Ted Pedersen, University of Minnesota Duluth
     tpederse at d.umn.edu

     Satanjeev Banerjee, Carnegie Mellon University
     banerjee+ at cs.cmu.edu

DOCUMENTATION COPYRIGHT AND LICENSE
    Copyright (C) 2005 Siddharth Patwardhan, Ted Pedersen, and Satanjeev
    Banerjee

    Permission is granted to copy, distribute and/or modify this document
    under the terms of the GNU Free Documentation License, Version 1.2 or
    any later version published by the Free Software Foundation; with no
    Invariant Sections, no Front-Cover Texts, and no Back-Cover Texts.

    Note: a copy of the GNU Free Documentation License is available on the
    web at <http://www.gnu.org/copyleft/fdl.html> and is included in this
    distribution as FDL.txt.



Changes
NAME
    Revision History for WordNet::SenseRelate::TargetWord.

SYNOPSIS
    A list of changes to the WordNet::SenseRelate::TargetWord package. These
    are copied from the Recently Completed Items in the file todo.pod when a
    new version is released.

DESCRIPTION
  Version 0.08 (Released 06/30/05)
    *   06/29/05

        (1) Changed the contextpos-based sense restriction process. (Sid)

        (2) Made some changes to README. (Sid)

    *   06/26/05

        (1) Added graphical interface senserelate.pl. (Sid)

        (2) Fixed bug in context selection module. (Sid)

  Version 0.07 (Released 06/24/05)
    *   06/23/05

        (1) Added graphical interface disamb-gui.pl. (Sid)

        (2) Added documentation. (Sid)

    *   06/19/05

        (1) Added output traces to all the modules (Context.pm,
            Compounds.pm, Global.pm, Local.pm, Random.pm, SenseOne.pm).
            (Sid)

        (2) Fixed bug in Local.pm. (Sid)

  Version 0.06 (Released 06/13/05)
    *   06/07/05

        (1) Converted the SenseRelate (v0.05) package to a modular package
            WordNet::SenseRelate::TargetWord. (Sid)

AUTHORS
      Siddharth Patwardhan, University of Utah, Salt Lake City
      sidd at cs.utah.edu

      Ted Pedersen, University of Minnesota Duluth
      tpederse at d.umn.edu

      Satanjeev Banerjee, Carnegie Mellon University, Pittsburgh
      banerjee+ at cs.cmu.edu

BUGS
    None.

SEE ALSO
    todo.pod

COPYRIGHT
    Copyright (c) 2005 Siddharth Patwardhan, Ted Pedersen and Satanjeev
    Banerjee

    Permission is granted to copy, distribute and/or modify this document
    under the terms of the GNU Free Documentation License, Version 1.2 or
    any later version published by the Free Software Foundation; with no
    Invariant Sections, no Front-Cover Texts, and no Back-Cover Texts.

    Note: a copy of the GNU Free Documentation License is available on the
    web at <http://www.gnu.org/copyleft/fdl.html> and is included in this
    distribution as FDL.txt.



utils/disamb.pl
#! /usr/local/bin/perl -w
# (Updated: $Id: disamb.pl,v 1.10 2005/06/24 13:55:39 sidz1979 Exp $)
#
# disamb.pl version 0.07
#
# Program that reads in a Senseval-2 formatted lexical sample file, and
# disambiguates the instances using WordNet::SenseRelate::TargetWord.
#
# Copyright (c) 2005
#
# Satanjeev Banerjee, Carnegie Mellon University
# banerjee+@cs.cmu.edu
#
# Ted Pedersen, University of Minnesota, Duluth
# tpederse@d.umn.edu
#
# Siddharth Patwardhan, University of Utah
# sidd@cs.utah.edu
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
#
#-----------------------------------------------------------------------------

# Uses:
use strict;
use WordNet::SenseRelate::Reader::Senseval2;
use WordNet::SenseRelate::TargetWord;
use WordNet::SenseRelate::Tools;
use Getopt::Long;

# Now get the options!
our ($opt_config, $opt_wnpath, $opt_trace, $opt_help, $opt_version);
&GetOptions("config=s", "wnpath=s", "trace", "help", "version");

# If help is requested
if (defined $opt_help)
{
    &printHelp();
    exit;
}

# If version information is requested
if (defined $opt_version)
{
    &printVersion();
    exit;
}

# Options
my $options = {};

# Load options if any
if (defined $opt_config)
{

    # Make all the options undefined first
    $options                      = {};
    $options->{preprocess}        = [];
    $options->{preprocessconfig}  = [];
    $options->{postprocess}       = [];
    $options->{postprocessconfig} = [];
    $options->{context}           = undef;
    $options->{contextconfig}     = undef;
    $options->{algorithm}         = undef;
    $options->{algorithmconfig}   = undef;
    $options->{wntools}           = undef;

    # Read the configuration file
    my $optStruct = &readConfigFile($opt_config);
    &askHelp("disamb.pl: Configuration file format error.\n")
      if (!defined $optStruct);

    # Now get the options from the options structure
    # First, the preprocess modules and options
    if (defined $optStruct->{'PREPROCESS'})
    {
        foreach my $element (@{$optStruct->{'PREPROCESS'}})
        {
            push(@{$options->{preprocess}}, $element->{'name'});
            delete $element->{'name'};
            push(@{$options->{preprocessconfig}}, $element);
        }
    }

    # Next, the postprocess modules and options
    if (defined $optStruct->{'POSTPROCESS'})
    {
        foreach my $element (@{$optStruct->{'POSTPROCESS'}})
        {
            push(@{$options->{postprocess}}, $element->{'name'});
            delete $element->{'name'};
            push(@{$options->{postprocessconfig}}, $element);
        }
    }

    # Next, get the context selection module and options
    if (defined $optStruct->{'CONTEXT'})
    {

        # There should be only one context selection module specified
        &askHelp(
"disamb.pl: Multiple context selection modules specified in configuration file.\n"
          )
          if (scalar(@{$optStruct->{'CONTEXT'}}) > 1);

        # Get the context selection module name, and options
        # If none are specified, the default will be used
        my $element = shift(@{$optStruct->{'CONTEXT'}});
        if (defined $element && defined $element->{'name'})
        {
            $options->{context} = $element->{'name'};
            delete $element->{'name'};
            $options->{contextconfig} = $element;
        }
    }

    # Next, get the algorithm module and options
    if (defined $optStruct->{'ALGORITHM'})
    {

        # Only one algorithm module must be specified
        &askHelp(
"disamb.pl: Multiple algorithm modules specified in configuration file.\n")
          if (scalar(@{$optStruct->{'ALGORITHM'}}) > 1);

        # Get the algorithm module name, and options
        # If none are specified, then the default will be used
        my $element = shift(@{$optStruct->{'ALGORITHM'}});
        if (defined $element && defined $element->{'name'})
        {
            $options->{algorithm} = $element->{'name'};
            delete $element->{'name'};
            $options->{algorithmconfig} = $element;
        }
    }
}

# Load WordNet::SenseRelate::Tools:
print STDERR "Loading WordNet::SenseRelate::Tools... ";
my $wntools = WordNet::SenseRelate::Tools->new($opt_wnpath);
&askHelp("disamb.pl: Unable to load WordNet::SenseRelate::Tools.\n")
  if (!defined $wntools);
print STDERR "done.\n";
$options->{'wntools'} = $wntools;

# Read Senseval2 data:
print STDERR "Reading XML data... ";
my $reader = WordNet::SenseRelate::Reader::Senseval2->new(shift);
&askHelp("disamb.pl: Unable to read XML file.\n") if (!defined $reader);
print STDERR "done.\n";

# Create WordNet::SenseRelate::TargetWord object:
my ($targetWord, $error) =
  WordNet::SenseRelate::TargetWord->new($options,
                                        ((defined $opt_trace) ? 1 : 0));
if (!defined $targetWord)
{
    print STDERR "$error\n";
    exit;
}

# Get each instance from the reader object, and disambiguate
foreach my $i (0 .. ($reader->instanceCount() - 1))
{
    my $instance = $reader->instance($i);
    my ($sense, $errString) = $targetWord->disambiguate($instance);
    if (!defined $sense && defined $errString && $errString ne "")
    {
        print STDERR "$errString\n";
        exit;
    }
    print STDERR $targetWord->getTraceString() if(defined $opt_trace);
    print $instance->{lexelt};
    print " ";
    print $instance->{id};
    print " $sense" if (defined $sense);
    print "\n";
}

# Subroutine that reads the options file
sub readConfigFile
{
    my $fname = shift;

    open(CONFIG, "$fname")
      || &askHelp("disamb.pl: Unable to open specified configuration file.\n");
    my $line = <CONFIG>;
    $line =~ s/[\r\f\n]//;
    $line =~ s/^\s+//;
    $line =~ s/\s+$//;
    &askHelp("disamb.pl: File format error (in header) in config file.\n")
      if ($line !~ /^WordNet::SenseRelate::TargetWord$/);
    my $struct  = {};
    my $section = "";
    my $modname = "";
    my $modata  = undef;

    while ($line = <CONFIG>)
    {
        $line =~ s/[\r\f\n]//;
        $line =~ s/\#.*//;
        $line =~ s/^\s+//;
        $line =~ s/\s+$//;
        if ($line =~ /\[SECTION:([^\]]*)\]/)
        {
            $section = $1;
            $struct->{$section} = []
              if (   defined $section
                  && $section ne ""
                  && !exists($struct->{$section}));
        }
        elsif ($line =~ /\[START\s+([^\]]+)\]/)
        {
            return undef if (!defined $section || $section eq "");
            $modname = $1;
            $modata  = {};
        }
        elsif ($line =~ /\[END\]/)
        {
            return undef
              if (   !defined $section
                  || $section eq ""
                  || !defined $modname
                  || $modname eq "");
            return undef if (!defined $modata);
            $modata->{'name'} = $modname;
            push(@{$struct->{$section}}, $modata);
            $modname = "";
            $modata  = undef;
        }
        elsif ($line =~ /([^=]+)\s*=\s*([^=]+)/)
        {
            return undef
              if (   !defined $section
                  || !defined $modname
                  || $section eq ""
                  || $modname eq "");
            $modata->{$1} = $2;
        }
        elsif ($line ne "")
        {
            return undef;
        }
    }
    close(CONFIG);

    return $struct;
}

# function to output help messages for this program
sub printHelp
{
    print
      "Performs Word Sense Disambiguation on Senseval-2 lexical sample data.\n";
    print
"Usage: disamb.pl [ [--config FILE] [--wnpath WNPATH] [--trace] XMLFILE | --help | --version]\n";
    print
      "--config         Specifies a configuration file (FILE) to set up the\n";
    print "                 various configuration options.\n";
    print
      "--wnpath         WNPATH specifies the path of the WordNet data files.\n";
    print
"                 Ordinarily, this path is determined from the \$WNHOME\n";
    print
      "                 environment variable. But this option overides this\n";
    print "                 behavior.\n";
    print "--trace          Indicates that trace information be printed.\n";
    print "--help           Displays this help screen.\n";
    print "--version        Displays version information.\n\n";
}

# function to output "ask for help" message when the user's goofed up!
sub askHelp
{
    my $message = shift;
    print STDERR "$message\n" if (defined $message && $message ne "");
    print STDERR "Type disamb.pl --help for help.\n";
    exit;
}

# Subroutine to print the version information
sub printVersion
{
    print "disamb.pl version 0.07\n";
    print
"Copyright (c) 2005 Satanjeev Banerjee, Ted Pedersen and Siddharth Patwardhan.\n";
}

__END__


=head1 NAME

disamb.pl - command line interface for WordNet::SenseRelate::TargetWord, a Word Sense Disambiguation 
module.

=head1 SYNOPSIS

disamb.pl [ [--config FILE] [--wnpath WNPATH] [--trace] XMLFILE | --help | --version ]

=head1 DESCRIPTION

This program is a command line interface to the WordNet::SenseRelate::TargetWord Word Sense Disambiguation
module. It takes as input a Senseval-2 formatted input file and disambiguated each instance specified in
file. The module is highly configurable. The user is allowed to choose different preprocessing and
postprocessing tasks for the data, the dismabiguation algorithm and the context selection algorithm.

=head1 OPTIONS

Usage: disamb.pl [ [--config FILE] [--wnpath WNPATH] [--trace] XMLFILE | --help | --version ]

B<--config>=I<FILE>         
    Specifies a configuration file (FILE) to set up the various configuration options.

B<--wnpath>=I<WNPATH>         
    WNPATH specifies the path of the WordNet data files. Ordinarily, this path is 
    determined from the $WNHOME environment variable. But this option overides this
    behavior.

B<--trace>          
    Indicates that trace information be printed.

B<--help>
    Displays this help screen.

B<--version>
    Displays version information.

=head1 AUTHORS

 Siddharth Patwardhan <sidd at cs.utah.edu>

 Ted Pedersen <tpederse at d.umn.edu>

 Satanjeev Banerjee <banerjee+ at cs.cmu.edu>

=head1 BUGS

=head1 SEE ALSO

perl(1)

WordNet::SenseRelate::TargetWord(3)

WordNet::Similarity(3)

http://www.cogsci.princeton.edu/~wn/

http://senserelate.sourceforge.net

http://groups.yahoo.com/group/senserelate/

=head1 COPYRIGHT

Copyright (c) 2005 Siddharth Patwardhan, Ted Pedersen, Satanjeev
Banerjee

This program is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation; either version 2 of the License, or (at your
option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

=cut
